plugin/src/main/java/io/deephaven/plugin/type/ObjectType.java

```diff
@@ -46,7 +47,85 @@ public interface ObjectType extends Plugin {
      * @param out the output stream
      * @throws IOException if an IO exception occurs
      */
-    void writeTo(Exporter exporter, Object object, OutputStream out) throws IOException;
+    default void writeTo(Exporter exporter, Object object, OutputStream out) throws IOException {
+        if (supportsBidiMessaging(object) == Kind.BIDIRECTIONAL) {
+            // internal error, shouldn't have called this
+            throw new IllegalStateException(
+                    "Do not call writeTo if supportsBidiMessaging returns true, but writeTo is not implemented");
+        } else {
+            // incorrect implementation
+            throw new IllegalStateException("ObjectType implementation returned false for supportsBidiMessaging");
+        }
+    }
+
+    /**
+     * A stream of messages, either sent from the server to the client, or client to the server. ObjectType plugin
+     * implementations can provide an implementation of this interface for each incoming stream to invoke as messages
+     * arrive, and will likewise be given an instance of this interface to be able to send messages to the client.
+     */
+    interface MessageStream extends AutoCloseable {
+        /**
+         * Transmits data to the remote end of the stream. This can consist of a binary payload and references to
+         * objects on the server.
+         *
+         * @param payload the binary data sent to the remote implementation
+         * @param references server-side object references sent to the remote implementation
+         */
+        void onMessage(ByteBuffer payload, Object[] references);
```

Comment:
We discussed that for outbound, this will export each of the references, regardless of duplication. This feels good to me, as it's more predictable. Maybe we want to (eventually) offer some kind of library support for references by index into the array, or with associated "naming", to make it clear how plugins should convey "the same server-side object is filling two roles here, in one message or across messages".

Reply:
Does this also mean that `io.deephaven.plugin.type.ObjectType.Exporter` goes away (and it becomes an implementation detail)?

I'm not opposed, and it may simplify things for fetch plugins as well. For example, the pandas plugin may go from something like:

```python
    def to_bytes(self, exporter: Exporter, data_frame: DataFrame):
        exporter.reference(to_table(data_frame))
        return b''
 ```
 
 to
  
```python
    def to_message(self, data_frame: DataFrame):
        return (b'', [ to_table(data_frame) ])
 ```
 
 or something like that (naming tbd).
 
 The plugins already are able to have their own format to do "reference by index"; https://github.com/deephaven/deephaven-plugin-json/blob/main/test/nested_node_types.json shows an example of this.

Reply:
Yes, the `Exporter` comes out of the API. Hooray, right?

Reply:
Exporter is still used in FetchOnly, not sure if I like how the classes are put together right now. It no longer is really part of ObjectType, but part of ObjectTypeBase.FetchOnly instead. Remove the interface and just make it a class, that keeps track of references/bytes? Remove rather than deprecate the unusable methods?